mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp

```diff
@@ -419,61 +420,104 @@ struct CustomCallOpPattern : public OpConversionPattern<CustomCallOp> {
     }
 };
 
-struct PythonCallOpPattern : public OpConversionPattern<PythonCallOp> {
+struct DefineCallbackOpPattern : public OpConversionPattern<CallbackOp> {
     using OpConversionPattern::OpConversionPattern;
 
-    LogicalResult matchAndRewrite(PythonCallOp op, PythonCallOpAdaptor adaptor,
-                                  ConversionPatternRewriter &rewriter) const override
+    LogicalResult match(CallbackOp op) const override
     {
-        MLIRContext *ctx = op.getContext();
-        Location loc = op.getLoc();
+        // Only match with ops without an entry block
+        return !op.empty() ? failure() : success();
+    }
 
-        // Create function
-        Type voidType = LLVM::LLVMVoidType::get(ctx);
-        auto point = rewriter.saveInsertionPoint();
-        ModuleOp mod = op->getParentOfType<ModuleOp>();
-        rewriter.setInsertionPointToStart(mod.getBody());
+    void rewrite(CallbackOp op, CallbackOpAdaptor adaptor,
+                 ConversionPatternRewriter &rewriter) const override
+    {
+        Block *entry;
+        rewriter.updateRootInPlace(op, [&] { entry = op.addEntryBlock(); });
+        PatternRewriter::InsertionGuard guard(rewriter);
+        rewriter.setInsertionPointToStart(entry);
+
+        auto ctx = rewriter.getContext();
+        Type ptrTy = LLVM::LLVMPointerType::get(ctx);
+        auto one = rewriter.getI64IntegerAttr(1);
+        auto loc = op.getLoc();
+        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, one);
+        auto idAttr = op.getIdAttr();
+        auto constantId = rewriter.create<LLVM::ConstantOp>(loc, idAttr);
+        auto argcAttr = op.getArgcAttr();
+        auto constantArgc = rewriter.create<LLVM::ConstantOp>(loc, argcAttr);
+        auto rescAttr = op.getRescAttr();
+        auto constantResc = rewriter.create<LLVM::ConstantOp>(loc, rescAttr);
+
+        SmallVector<Value> callArgs = {constantId, constantArgc, constantResc};
 
         Type i64 = rewriter.getI64Type();
-        // The argument convention is as follows:
-        // arg0 =        identifier
-        // arg1 =        length of operands
-        // arg2 =        length of results
-        // arg3..N+3     varargs pointers to memrefs
-        // argN+4..N+M+4 varargs pointers to result memrefs
-
+        Type voidType = LLVM::LLVMVoidType::get(ctx);
         bool isVarArg = true;
-        LLVM::LLVMFuncOp customCallFnOp = mlir::LLVM::lookupOrCreateFn(
-            mod, "inactive_callback", {/*args=*/i64, i64, i64}, /*ret_type=*/voidType, isVarArg);
-        customCallFnOp.setPrivate();
-        rewriter.restoreInsertionPoint(point);
+        ModuleOp mod = op->getParentOfType<ModuleOp>();
+        auto typeConverter = getTypeConverter();
+        LLVM::LLVMFuncOp customCallFnOp =
+            mlir::LLVM::lookupOrCreateFn(mod, "inactive_callback", {/*args=*/i64, i64, i64},
+                                         /*ret_type=*/voidType, isVarArg);
+
+        for (auto arg : op.getArguments()) {
```

Comment:
```suggestion
        // TODO: remove redundant alloca+store since ultimately we'll receive struct*
        for (auto arg : op.getArguments()) {
```